### PR TITLE
Documentation harness introduced

### PR DESCRIPTION
**Pull request by VasilyStrelyaev**
To preview documentation locally, you'll have to install Jekyll first - http://jekyll-windows.juthilo.com/ for Windows.
Then install the redcarpet gem via `gem install redcarpet`
Finally, use `gulp preview-website`.

\cc @inikulin @AlexanderMoskovkin @churkin @kirovboris @MargaritaLoseva

### CODE REVIEW

**Comment by VasilyStrelyaev** [comment]
@testcafe-build-bot \retest

File: Gulpfile.js [hunk 1]
```diff
@@ -17,7 +19,16 @@ var publish      = require('publish-please');
...
 var path         = require('path');
+var opn          = require('opn');
+var connect      = require('connect');
+var readline     = require('readline');
+var spawn        = require('child_process').spawn;
+var serveStatic  = require('serve-static');
 var Promise      = require('pinkie');
+var markdownlint = require('markdownlint');
+// TODO: Uncomment when preparing the deploy-to-gh-pages task
+// var ghPages      = require('gulp-gh-pages');
```
Comment (inikulin) [inline Gulpfile.js]: This will be implemented in the separate PR, right?

File: Gulpfile.js [hunk 2]
```diff
@@ -265,6 +278,124 @@ gulp.task('test-client-travis', ['build'], function () {
...
+
+                if (lintErr)
+                    reject(lintErr);
+                else
+                    resolve();
+            });
+        });
+    }
+
+    return globby('./docs/articles/**/*.md').then(lintFiles);
```
Comment (inikulin) [inline Gulpfile.js]: I guess we should lint any `*.md` files that present in repo, e.g. README.md

File: Gulpfile.js [hunk 3]
```diff
@@ -265,6 +278,124 @@ gulp.task('test-client-travis', ['build'], function () {
...
+});
+
+gulp.task('clear-site', ['lint-docs'], function (cb) {
```
Comment (inikulin) [inline Gulpfile.js]: Why it's rely on linting?

File: Gulpfile.js (hunk 2 above)
Comment (inikulin) [inline Gulpfile.js]: Nevermind.

File: .markdownlint.json [hunk 4]
```diff
@@ -0,0 +1,8 @@
+{
```
Comment (inikulin) [inline .markdownlint.json]: Let's put this config file in `docs`

File: Gulpfile.js [hunk 5]
```diff
@@ -265,6 +278,124 @@ gulp.task('test-client-travis', ['build'], function () {
...
+        });
+        rl.write('Server stopped.\r\n');
+        sockets.forEach(function (socket) {
+            socket.destroy();
+        });
+        rl.close();
+    });
+});
+
+gulp.task('preview-website', ['prompt-web-server-close']);
```
Comment (inikulin) [inline Gulpfile.js]: Why we need `prompt-web-server-close` task since it's all the same as `preview-website`?

File: Gulpfile.js [hunk 6]
```diff
@@ -265,6 +278,124 @@ gulp.task('test-client-travis', ['build'], function () {
...
+            sockets.splice(sockets.indexOf(socket), 1);
+        });
+    });
+});
+
+gulp.task('open-website', ['serve-website'], function () {
+    opn('http://localhost:8080/testcafe');
+});
+
+gulp.task('prompt-web-server-close', ['open-website'], function (cb) {
```
Comment (inikulin) [inline Gulpfile.js]: No need for this task at all. I can just press `Ctrl-C` in my terminal to terminate server

File: Gulpfile.js [hunk 7]
```diff
@@ -265,6 +278,124 @@ gulp.task('test-client-travis', ['build'], function () {
...
+
+    cp.on('exit', function (code) {
+        cb(code === 0 ? null : 'ERROR: Jekyll process exited with code: ' + code);
+    });
+});
+
+gulp.task('serve-website', ['build-website'], function () {
+    server = connect().use('/testcafe', serveStatic('./site/deploy')).listen(8080);
+    server.on('connection', function (socket) {
+        sockets.push(socket);
```
Comment (inikulin) [inline Gulpfile.js]: No need to store this sockets

File: Gulpfile.js [hunk 8]
```diff
@@ -265,6 +278,124 @@ gulp.task('test-client-travis', ['build'], function () {
...
+        socket.on('close', function () {
+            sockets.splice(sockets.indexOf(socket), 1);
+        });
+    });
+});
+
+gulp.task('open-website', ['serve-website'], function () {
+    opn('http://localhost:8080/testcafe');
```
Comment (inikulin) [inline Gulpfile.js]: `server-website` resolves immediately without waiting for `listening` event. So, in theory, page can be opened before it can be served.

File: Gulpfile.js [hunk 9]
```diff
@@ -265,6 +278,124 @@ gulp.task('test-client-travis', ['build'], function () {
...
+    return gulp.src('./docs/_includes/**/*')
+		.pipe(gulp.dest('./site/src/_includes'));
+});
+
+gulp.task('put-in-docs', ['put-in-articles', 'put-in-navigation', 'put-in-templates']);
+
+gulp.task('prepare-website', ['fetch-assets-repo', 'put-in-docs']);
+
+gulp.task('build-website', ['prepare-website'], function (cb) {
+    var cp = spawn('jekyll.bat', ['build', '--source', './site/src/', '--destination', './site/deploy']);
```
Comment (inikulin) [inline Gulpfile.js]: What is `jekyll.bat`? This is windows-specific. So, I will not be able to run docs site on my Mac

File: Gulpfile.js [hunk 10]
```diff
@@ -265,6 +278,124 @@ gulp.task('test-client-travis', ['build'], function () {
...
+        cb();
+    });
+});
+
+gulp.task('put-in-articles', ['fetch-assets-repo'], function () {
+    return gulp.src('./docs/articles/**/*')
+		.pipe(gulp.dest('./site/src'));
+});
+
+gulp.task('put-in-navigation', ['put-in-articles'], function () {
```
Comment (inikulin) [inline Gulpfile.js]: Why it depends on `put-in-articles`? Can't it be done in parallel?

File: Gulpfile.js [hunk 11]
```diff
@@ -265,6 +278,124 @@ gulp.task('test-client-travis', ['build'], function () {
...
+    }
+
+    return globby('./docs/articles/**/*.md').then(lintFiles);
+});
+
+gulp.task('clear-site', ['lint-docs'], function (cb) {
+    del('./site/**', function (e) {
+        cb(e);
+    });
+});
```
Comment (inikulin) [inline Gulpfile.js]: No need for closure, just `cb` directly to `del`

File: Gulpfile.js [hunk 12]
```diff
@@ -265,6 +278,124 @@ gulp.task('test-client-travis', ['build'], function () {
...
+    opn('http://localhost:8080/testcafe');
+});
+
+gulp.task('prompt-web-server-close', ['open-website'], function (cb) {
+    var rl = readline.createInterface({
+        input:  process.stdin,
+        output: process.stdout
+    });
+
+    rl.question('Press Enter to terminate the web server...', function () {
```
Comment (inikulin) [inline Gulpfile.js]: For `publish` I suggest you to use https://github.com/SBoudrias/Inquirer.js/ instead of `readline`

File: Gulpfile.js (hunk 1 above)
Comment (VasilyStrelyaev) [inline Gulpfile.js]: yeah, after the docs is ready so that it's safe to deploy

File: Gulpfile.js [hunk 13]
```diff
@@ -265,6 +278,124 @@ gulp.task('test-client-travis', ['build'], function () {
...
+                    resolve();
+            });
+        });
+    }
+
+    return globby('./docs/articles/**/*.md').then(lintFiles);
+});
+
+gulp.task('clear-site', ['lint-docs'], function (cb) {
+    del('./site/**', function (e) {
```
Comment (inikulin) [inline Gulpfile.js]: No need for relative paths in gulpfile and for globs here as well. Just `del('stie', cb); See `clean` implementation above. Also, maybe we should rename it to `clean-site` for consistency. Also, some tasks use term `website` others just `site`, dir is named `site` as well. We should end up we one common term. I prefer `website`

File: Gulpfile.js (hunk 1 above)
Comment (inikulin) [inline Gulpfile.js]: Maybe we just remove this dead code then for now and create a GH issue for this functionality?

**Comment by inikulin** [comment]
\r-

File: .markdownlint.json (hunk 4 above)
Comment (VasilyStrelyaev) [inline .markdownlint.json]: VS Code looks for it in the root.
Of course, I can live without VS Code highlighting rule violations - if this file is messy here.

Comment (inikulin) [inline .markdownlint.json]: Let's keep here then

File: Gulpfile.js (hunk 3 above)
Comment (VasilyStrelyaev) [inline Gulpfile.js]: I was thinking in a different way: we have nothing to do until docs are not okay, So, lint is executed at first, without a parallel task.

File: Gulpfile.js (hunk 6 above)
Comment (VasilyStrelyaev) [inline Gulpfile.js]: Cool! Didn't know that!

File: Gulpfile.js (hunk 1 above)
Comment (VasilyStrelyaev) [inline Gulpfile.js]: I was afraid of losing working code... 
But you're right, I'll keep it somewhere else so far

File: Gulpfile.js (hunk 7 above)
Comment (VasilyStrelyaev) [inline Gulpfile.js]: As far as we now stop the server with `Ctrl-C` - yeah, no need

File: Gulpfile.js (hunk 5 above)
Comment (VasilyStrelyaev) [inline Gulpfile.js]: No longer need - thanks to `Ctrl-C`

File: Gulpfile.js (hunk 10 above)
Comment (VasilyStrelyaev) [inline Gulpfile.js]: That's what I tried to do at first, but got some strange mess and resorted to sequential execution.
Should have investigated it, sure...

File: Gulpfile.js (hunk 1 above)
Comment (inikulin) [inline Gulpfile.js]: You can drop snippet in the issue

File: Gulpfile.js (hunk 9 above)
Comment (VasilyStrelyaev) [inline Gulpfile.js]: Ooops. Forgot about Macs. Need to use `jekyll` for Mac and `jekyll.bat` for Windows

Comment (VasilyStrelyaev) [inline Gulpfile.js]: `jekyll` doesn't work for Windows with `spawn`
http://stackoverflow.com/questions/17516772/using-nodejss-spawn-causes-unknown-option-and-error-spawn-enoent-err/17537559#17537559

Comment (inikulin) [inline Gulpfile.js]: Use https://www.npmjs.com/package/cross-spawn-async module then

Comment (inikulin) [inline Gulpfile.js]: BTW, you should set `{ stdio: 'inherit' }` option I guess. So the output of the `jekyll` will be shown in terminal

**Comment by inikulin** [comment]
@testcafe-build-bot \retest

File: Gulpfile.js [hunk 14]
```diff
@@ -265,6 +273,64 @@ gulp.task('test-client-travis', ['build'], function () {
...
+    return gulp.src('docs/articles/**/*')
+		.pipe(gulp.dest('website/src'));
+});
+
+gulp.task('put-in-navigation', ['fetch-assets-repo'], function () {
+    return gulp.src('docs/_data/**/*')
+		.pipe(gulp.dest('website/src/_data'));
+});
+
+gulp.task('prepare-website', ['fetch-assets-repo', 'put-in-articles', 'put-in-navigation', 'lint-docs']);
```
Comment (inikulin) [inline Gulpfile.js]: I think it shouldn't rely on `fetch-assets-repo`, since other tasks already rely on it

File: Gulpfile.js [hunk 15]
```diff
@@ -265,6 +273,64 @@ gulp.task('test-client-travis', ['build'], function () {
...
+
+gulp.task('build-website', ['prepare-website'], function (cb) {
+    spawn('jekyll', ['build', '--source', 'website/src/', '--destination', 'website/deploy'], { stdio: 'inherit' })
+        .on('exit', cb);
+});
+
+gulp.task('serve-website', ['build-website'], function (cb) {
+    connect()
+        .use('/testcafe', serveStatic('website/deploy'))
+        .listen(8080, cb);
```
Comment (inikulin) [inline Gulpfile.js]: We can combine `server-website` and `preview-website` into the one task. Since I don't see usage scenarios for the `server-website` alone

**Comment by inikulin** [comment]
\r-

File: Gulpfile.js [hunk 16]
```diff
@@ -265,6 +273,64 @@ gulp.task('test-client-travis', ['build'], function () {
...
+    git.clone('https://github.com/VasilyStrelyaev/testcafe-gh-page-assets.git', { args: 'website' }, cb);
+});
+
+gulp.task('put-in-articles', ['fetch-assets-repo'], function () {
+    return gulp.src('docs/articles/**/*')
+		.pipe(gulp.dest('website/src'));
+});
+
+gulp.task('put-in-navigation', ['fetch-assets-repo'], function () {
+    return gulp.src('docs/_data/**/*')
```
Comment (inikulin) [inline Gulpfile.js]: Do we really need to follow jekyll's weird naming here. I think it will be better to keep this in `docs/nav` instead of `docs/_data`. Then on build step we can copy it to the appropriate dir.

File: Gulpfile.js [hunk 17]
```diff
@@ -265,6 +273,64 @@ gulp.task('test-client-travis', ['build'], function () {
...
+gulp.task('clean-website', function (cb) {
+    del('website', cb);
+});
+
+gulp.task('fetch-assets-repo', ['clean-website'], function (cb) {
+    git.clone('https://github.com/VasilyStrelyaev/testcafe-gh-page-assets.git', { args: 'website' }, cb);
+});
+
+gulp.task('put-in-articles', ['fetch-assets-repo'], function () {
+    return gulp.src('docs/articles/**/*')
```
Comment (inikulin) [inline Gulpfile.js]: Here and in all other place: newline after `gulp`

**Comment by inikulin** [comment]
\r-

File: Gulpfile.js (hunk 14 above)
Comment (VasilyStrelyaev) [inline Gulpfile.js]: yeah, I know, I added it to make the `prepare-website` task describe what this preparation consists of.

Comment (inikulin) [inline Gulpfile.js]: No need for this. The rule is that gulp deps chain should be as clean as possible. It's much easier to understand that's going on this way.

**Comment by VasilyStrelyaev** [comment]
@testcafe-build-bot \retest

**Comment by VasilyStrelyaev** [comment]
FPR

**Comment by inikulin** [comment]
Awesome, lgtm.

**Comment by VasilyStrelyaev** [comment]
Will anyone else have a say, or just merge?

**Comment by kirovboris** [comment]
lgtm

**Comment by inikulin** [comment]
ping @AlexanderMoskovkin @MargaritaLoseva @churkin

**Comment by AlexanderMoskovkin** [comment]
lgtm

**Comment by inikulin** [comment]
@VasilyStrelyaev I guess should transfer https://github.com/VasilyStrelyaev/testcafe-gh-page-assets before merging, so you'll be able to correct URL in context of current PR.

**Comment by inikulin** [comment]
@VasilyStrelyaev Oh, transfer email has expired. BTW, why you transfering it to me? You can add me to project collaborators so I'll be able to transfer it directly to the DevExpress org

**Comment by VasilyStrelyaev** [comment]
@inikulin  ok, let's try

**Comment by VasilyStrelyaev** [comment]
I'd be surprised if collaborators could transfer repositories

**Comment by inikulin** [comment]
@VasilyStrelyaev nevermind, just promoted you to owners in DevExpress org. You can transfer on your own now.

**Comment by VasilyStrelyaev** [comment]
@inikulin thx, transfered

**Comment by inikulin** [comment]
@testcafe-build-bot \retest

**Comment by VasilyStrelyaev** [comment]
@testcafe-build-bot \retest

**Comment by AlexanderMoskovkin** [comment]
FYI: Guys, now the tests are failed because of  `saucelabs environment error`, this is an error not on our side. Saucelabs often can't run OSX virtual machines for some reason (I don't know why). We with @georgiy-abbasov are working on improving `qunit-harness` to stable this

**Comment by AlexanderMoskovkin** [comment]
@testcafe-build-bot \retest

**Comment by AlexanderMoskovkin** [comment]
@testcafe-build-bot \retest

**Comment by inikulin** [comment]
@testcafe-build-bot \retest
